### PR TITLE
Feature/plugins

### PR DESCRIPTION
**Pull request by mmoreram**

### CODE REVIEW

**Comment by mmoreram** [comment]
Related with:

Bamboo - https://github.com/elcodi/bamboo/pull/26
StoreTemplateBundle - https://github.com/elcodi-templates/StoreTemplateBundle/pull/6
StoreWhiteTemplateBundle - https://github.com/elcodi-templates/StoreWhiteTemplateBundle/pull/3

File: src/Elcodi/Bundle/BambooBundle/DependencyInjection/ElcodiBambooExtension.php [hunk 1]
```diff
@@ -109,10 +107,14 @@ protected function getParametrizationValues(array $config)
      */
     public function getConfigFiles(array $config)
     {
+        $bambooInstalled = class_exists('\AppKernel');
+
+        if (!$bambooInstalled) {
```
Comment (xphere) [inline src/Elcodi/Bundle/BambooBundle/DependencyInjection/ElcodiBambooExtension.php]: What's the rationale behind that?

File: src/Elcodi/Component/Configuration/Services/ConfigurationManager.php [hunk 2]
```diff
@@ -186,13 +187,25 @@ public function get($configurationIdentifier)
 
         if (!($configurationLoaded instanceof ConfigurationInterface)) {
 
-            $parameterReference = $this->configurationElements[$configurationIdentifier]['reference'];
+            $configurationElement = $this->configurationElements[$configurationIdentifier];
+            $configurationValue = isset($configurationElement['reference'])
+                ? $this->parameterBag->get($configurationElement['reference'])
+                : $configurationElement['default_value'];
+
+            if (empty($configurationValue) && !$configurationElement['can_be_empty']) {
```
Comment (xphere) [inline src/Elcodi/Component/Configuration/Services/ConfigurationManager.php]: May be better to check with `"" === $configuration`?

Comment (mmoreram) [inline src/Elcodi/Component/Configuration/Services/ConfigurationManager.php]: Well, can be "", can be null...

File: src/Elcodi/Component/Plugin/Services/PluginManager.php [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                        'version'             => 'Any',
+                        'author'              => 'Anonymous',
+                        'year'                => 'NaN',
+                        'url'                 => '',
+                        'fa_icon'             => 'gear',
+                        'configuration_route' => null,
+                        'enabled'             => false,
+                        'configuration'       => []
+                    ],
+                    $this->getPluginSpecification($bundle->getPath())
```
Comment (xphere) [inline src/Elcodi/Component/Plugin/Services/PluginManager.php]: Maybe we should not allow the plugin to override some keys? Like `bundle` or `namespace`.

File: src/Elcodi/Component/Plugin/Services/PluginManager.php [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        $yaml = new Parser();
+        $specificationFilePath = $bundlePath . '/plugin.yml';
+
+        if (!file_exists($specificationFilePath)) {
+            return [];
+        }
+
+        return array_intersect_key(
+            $yaml->parse(file_get_contents($specificationFilePath)),
+            array_flip([
```
Comment (xphere) [inline src/Elcodi/Component/Plugin/Services/PluginManager.php]: Oops, sorry, remove last comment. I :heart: whitelists.

File: src/Elcodi/Component/Plugin/Services/PluginManager.php (hunk 3 above)
Comment (mmoreram) [inline src/Elcodi/Component/Plugin/Services/PluginManager.php]: And we cannot. You will see the `array_intersect_key` method in getPluginSpecification

File: src/Elcodi/Plugin/PinterestBundle/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+Elcodi Banner Bundle for Symfony2
```
Comment (xphere) [inline src/Elcodi/Plugin/PinterestBundle/README.md]: Not BannerBundle anymore.

File: src/Elcodi/Plugin/PinterestBundle/Templating/TwigRenderer.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * Renders import/export buttons
+     *
+     * @param EventInterface $event Event
+     */
+    public function renderJavascript(EventInterface $event)
+    {
+        if (
+            !$this->plugin->isEnabled() ||
```
Comment (xphere) [inline src/Elcodi/Plugin/PinterestBundle/Templating/TwigRenderer.php]: I'd do this check in an upper layer and don't let the plugin do anything while disabled.

File: src/Elcodi/Plugin/PinterestBundle/composer.json [hunk 7]
```diff
@@ -0,0 +1,52 @@
+{
+    "name": "elcodi-plugin/pinterest",
+    "description": "Elcodi Pinterest plugin,
```
Comment (xphere) [inline src/Elcodi/Plugin/PinterestBundle/composer.json]: Missing `"` before the `,`

File: src/Elcodi/Plugin/PinterestBundle/Templating/TwigRenderer.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Renders asynchronous javascript
+     *
+     * @param EventInterface $event Event
+     */
+    public function renderAsynchronousJavascript(EventInterface $event)
+    {
+        if (
+            !$this->plugin->isEnabled() ||
+            !isset($this->plugin->getConfiguration()['asynchronous']) ||
```
Comment (xphere) [inline src/Elcodi/Plugin/PinterestBundle/Templating/TwigRenderer.php]: If all we need is to access the plugin configuration, maybe we should flip to a this dependency, not the whole plugin.

File: src/Elcodi/Component/Configuration/Services/ConfigurationManager.php (hunk 2 above)
Comment (xphere) [inline src/Elcodi/Component/Configuration/Services/ConfigurationManager.php]: I'd better check for `""` and `null` then.

The magic behind `empty` is scary.

magic

File: src/Elcodi/Plugin/PinterestBundle/composer.json [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    ],
+    "require": {
+        "php": ">=5.4",
+
+        "symfony/dependency-injection": "~2.6",
+        "symfony/http-kernel": "~2.6",
+
+        "elcodi/plugin-bundle": "~0.4.0",
+        "elcodi/"
```
Comment (xphere) [inline src/Elcodi/Plugin/PinterestBundle/composer.json]: Doh! This dependency list is not even finished :sweat:

**Pull request merged**